contracts/staking/contracts/src/fees/MixinExchangeFees.sol

```diff
@@ -95,15 +99,39 @@ contract MixinExchangeFees is
         payable
         onlyExchange
     {
-        uint256 amount = msg.value;
+        // If the protocol fee payment is invalid, revert with a rich error.
+        if (
+            protocolFeePaid == 0 ||
+            (msg.value != protocolFeePaid && msg.value != 0)
+        ) {
+            LibRichErrors.rrevert(LibStakingRichErrors.InvalidProtocolFeePaymentError(
+                protocolFeePaid,
+                msg.value
+            ));
+        }
+
+        // Transfer the protocol fee to this address if it should be paid in WETH.
+        if (msg.value == 0) {
+            erc20Proxy.transferFrom(
```

Comment:
This comment is more directed to architecture opposed to the logic, but I'm wondering if we should deploy a separate `feeProxy`.

(i)
When users set their allowance on the ERC20 proxy they permit our exchange contract to transfer tokens on their behalf. The user then explicitly consents to the transfer of token X in amount Y when they sign an order.

Here we're overloading the ERC20 proxy to collect a hidden or implicit fee. We can broadcast to our community that there now exists a fee, but it's still a passive agreement that the average user may not understand. 

(ii)
The ERC20 proxy is super OP. It can transfer any ERC20 that the user has set allowance for. In this code, we really only need access to WETH.

(iii) 
This also raises a question around the proxy used to stake ZRX. On one hand, this doesn't break (i) because the user still explicitly agrees to the transfer by calling `stake()`. However (ii) does apply because in that scenario we really only need access to ZRX.

I propose we have either (i) a `wethFeeProxy` and a `zrxStakeProxy`, or (ii) a combined `stakingProxy` that users can set allowances for WETH and ZRX on. Note that this can just be the generic ERC20 Proxy, but redeployed.

@abandeali1 curious to get your thoughts.

Reply:
I have mixed feeling on it. I think I'm leaning towards having a separate `wethAssetProxy` but continuing to use the current `ERC20Proxy` for ZRX.

The good news is that we don't need to decide now since this can literally be a fork of the current `ERC20Proxy`. Let's just rename this variable `wethAssetProxy` (might as well rename ZRXVault.erc20Proxy to `zrxAssetProxy` while we're at it).